File: src/ios/FCMPlugin.m
#include <sys/types.h>
#include <sys/sysctl.h>

#import "AppDelegate+FCMPlugin.h"

#import <Cordova/CDV.h>
#import "FCMPlugin.h"
#import "Firebase.h"

@interface FCMPlugin () {}
@end

@implementation FCMPlugin

static BOOL notificatorReceptorReady = NO;
static BOOL appInForeground = YES;

static NSString *notificationCallback = @"FCMPlugin.onNotificationReceived";
static NSString *FCMIDCallback = @"FCMPlugin.onFCMIDReceived";

static FCMPlugin *fcmPluginInstance;

+ (FCMPlugin *) fcmPlugin {
    
    return fcmPluginInstance;
}

- (void) ready:(CDVInvokedUrlCommand *)command
{
    NSLog(@"Cordova view ready");
    fcmPluginInstance = self;
    [self.commandDelegate runInBackground:^{
        
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
    
}

// UN/SUBSCRIBE TOPIC //
- (void) subscribeToTopic:(CDVInvokedUrlCommand *)command 
{
    NSString* topic = [command.arguments objectAtIndex:0];
    NSLog(@"subscribe To Topic %@", topic);
    [self.commandDelegate runInBackground:^{
        if(topic != nil)[[FIRMessaging messaging] subscribeToTopic:[NSString stringWithFormat:@"/topics/%@", topic]];
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:topic];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
    
}

- (void) unsubscribeFromTopic:(CDVInvokedUrlCommand *)command 
{
    NSString* topic = [command.arguments objectAtIndex:0];
    NSLog(@"subscribe To Topic %@", topic);
    [self.commandDelegate runInBackground:^{
        if(topic != nil)[[FIRMessaging messaging] unsubscribeFromTopic:[NSString stringWithFormat:@"/topics/%@", topic]];
        CDVPluginResult* pluginResult = nil;
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:topic];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
    
}

- (void) setBadgeNumber:(CDVInvokedUrlCommand *)command {
    int number = [[command.arguments objectAtIndex:0] intValue];

    [self.commandDelegate runInBackground:^{
        [[UIApplication sharedApplication] setApplicationIconBadgeNumber:number];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) getBadgeNumber:(CDVInvokedUrlCommand *)command {
    [self.commandDelegate runInBackground:^{
        long badge = [[UIApplication sharedApplication] applicationIconBadgeNumber];

        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDouble:badge];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }];
}

- (void) registerNotification:(CDVInvokedUrlCommand *)command
{
    NSLog(@"view registered for notifications");
    
    notificatorReceptorReady = YES;
    NSData* lastPush = [AppDelegate getLastPush];
    if (lastPush != nil) {
        [FCMPlugin.fcmPlugin notifyOfMessage:lastPush];
    }
    
    CDVPluginResult* pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

//edited above 1
- (void) getFCMID:(CDVInvokedUrlCommand *)command
{
    NSLog(@"view registered for FCM ID");
    
    notificatorReceptorReady = YES;
    NSString* lastFCMID = [AppDelegate getLastFCMID];
    NSLog(@"trackFCM lastFCMID %@", lastFCMID);

    if (lastFCMID != nil) {
        [FCMPlugin.fcmPlugin notifyOfFCMID:lastFCMID];
    }
    
    CDVPluginResult* pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    
}

-(void) notifyOfMessage:(NSData *)payload
{
    NSString *JSONString = [[NSString alloc] initWithBytes:[payload bytes] length:[payload length] encoding:NSUTF8StringEncoding];
    NSString * notifyJS = [NSString stringWithFormat:@"%@(%@);", notificationCallback, JSONString]; 

    NSLog(@"trackFCM stringByEvaluatingJavaScriptFromString %@", notifyJS);
    
    if ([self.webView respondsToSelector:@selector(stringByEvaluatingJavaScriptFromString:)]) {
        [(UIWebView *)self.webView stringByEvaluatingJavaScriptFromString:notifyJS];
    } else {
        [self.webViewEngine evaluateJavaScript:notifyJS completionHandler:nil];
    }
}

//edited above 1

-(void) notifyOfFCMID:(NSString *)FCMID
{
    NSString * notifyJS = [NSString stringWithFormat:@"%@('%@');", FCMIDCallback, FCMID];
    NSLog(@"trackFCM stringByEvaluatingJavaScriptFromString %@", notifyJS);
    
    if (FCMID != nil) {
         NSLog(@"trackFCM if FCMID %@", FCMID);
        if ([self.webView respondsToSelector:@selector(stringByEvaluatingJavaScriptFromString:)]) {
            NSLog(@"trackFCM if in send to js");
            [(UIWebView *)self.webView stringByEvaluatingJavaScriptFromString:notifyJS];
        } else {
             NSLog(@"trackFCM else in send to js");
            [self.webViewEngine evaluateJavaScript:notifyJS completionHandler:nil];
        }
    }else{
         NSLog(@"trackFCM else FCMID %@", FCMID);
    }
}

-(void) appEnterBackground
{
    NSLog(@"Set state background");
    appInForeground = NO;
}

-(void) appEnterForeground
{
    NSLog(@"Set state foreground");
    NSData* lastPush = [AppDelegate getLastPush];
    if (lastPush != nil) {
        [FCMPlugin.fcmPlugin notifyOfMessage:lastPush];
    }
    appInForeground = YES;
}

@end

File: www/FCMPlugin.js
var exec = require('cordova/exec');

function FCMPlugin() {
	console.log("FCMPlugin.js: is created");
}

// GET TOKEN //
FCMPlugin.prototype.getToken = function (success, error) {
	console.log(124)
}

// // SUBSCRIBE TO TOPIC //
// FCMPlugin.prototype.subscribeToTopic = function (topic, success, error) {
// 	exec(success, error, "FCMPlugin", 'subscribeToTopic', [topic]);
// }
// // UNSUBSCRIBE FROM TOPIC //
// FCMPlugin.prototype.unsubscribeFromTopic = function (topic, success, error) {
// 	exec(success, error, "FCMPlugin", 'unsubscribeFromTopic', [topic]);
// }

// // NOTIFICATION CALLBACK //
// FCMPlugin.prototype.onNotification = function (callback, success, error) {
// 	FCMPlugin.prototype.onNotificationReceived = callback;
// 	exec(success, error, "FCMPlugin", 'registerNotification', []);
// }
// // DEFAULT NOTIFICATION CALLBACK //
// FCMPlugin.prototype.onNotificationReceived = function (payload) {
// 	console.log("fcmTrack Received push notification payload:")
// 	console.log(payload)
// }
// //Edit start wed june 22
// // FCM ID CALLBACK //
// FCMPlugin.prototype.onGotFCMID = function (callback, success, error) {
// 	FCMPlugin.prototype.onFCMIDReceived = callback;
// 	exec(success, error, "FCMPlugin", 'getFCMID', []);
// }
// // DEFAULT ID CALLBACK //
// FCMPlugin.prototype.onFCMIDReceived = function (fcmID) {
// 	console.log("fcmTrack Received fcmID:")
// 	console.log(fcmID)
// }
// //Edit end

// FCMPlugin.prototype.setBadgeNumber = function (number, success, error) {
// 	exec(success, error, "FCMPlugin", "setBadgeNumber", [number]);
// };

// FCMPlugin.prototype.getBadgeNumber = function (success, error) {
// 	exec(success, error, "FCMPlugin", "getBadgeNumber", []);
// };

// FIRE READY //
//exec(function (result) { console.log("fcmTrack FCMPlugin Ready OK") }, function (result) { console.log("fcmTrack FCMPlugin Ready ERROR") }, "FCMPlugin", 'ready', []);

var fcmPlugin = new FCMPlugin();
module.exports = fcmPlugin;